xarray/backends/api.py

```diff
@@ -764,10 +776,17 @@ def save_mfdataset(datasets, paths, mode='w', format=None, groups=None,
                          'save_mfdataset')
 
     writer = ArrayWriter()
-    stores = [to_netcdf(ds, path, mode, format, group, engine, writer)
+    stores = [to_netcdf(ds, path, mode, format, group, engine, writer,
+                        compute=compute)
               for ds, path, group in zip(datasets, paths, groups)]
+
     try:
-        writer.sync()
+        delayed = writer.sync(compute=compute)
+        if not compute:
+            import dask
+            # or should this return an iterable of delayed objects
+            # (one for each store)
+            return dask.delayed(_finalize_stores)(delayed, *stores)
```

Comment:
Thinking about this a little more, this is a sub-optimal way to setup the task graph for dask:
- Currently, you have: all write tasks -> close all files -> done
- Rather, we should have: set of all {write one file -> close the file}  -> done

The later is far less likely to run into issues with too many open files, because it closes each file immediately when it's done writing.

To do this, I think we should setup `to_netcdf()` to return delayed objects rather than stores if `compute=False`, in which case we can simply collect them all together in this result task.

We might also make dask required for `save_mfdataset` since it really is entirely useless otherwise and that would let us simplify this logic (e.g., to get ride of the `Writer` class).

Reply:
@shoyer - I think this is a good idea. However, I'm at a loss as to how to implement it. Should we just call `_finalize_store` in `to_netcdf` or should that happen after `writer.sync(compute=compute)` is called?